dialogue-core/src/main/java/com/palantir/dialogue/core/NodeSelectingChannel.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.palantir.dialogue.Request;
+
+interface NodeSelectingChannel extends LimitedChannel {
+    void routeToHost(int index, Request request);
```

Comment:
What if we used an API along the lines of:
```suggestion
    /** Returns a stable list of all known delegate nodes. */
    ImmutableList<LimitedChannel> nodeChannels();
```

Reply:
Sorry, I think the API would actually look like this:

```suggestion
    Refreshable<? extends List<? extends LimitedChannel> nodeChannels();
```

The refreshable allows us to update the channels with the current node-selection-strategy info as the strategy changes, however within a single DialogueChannel the order would be guaranteed to be stable.
The refreshable would also update on config reload which may or may not impact the list order.

Reply:
Done, looks good.

Reply:
Hmmm I'm not sure what the refreshable buys us?

1. The order and set of uris is always fixed within a DialogueChannel; it gets Config not Refreshable<Config>. Refreshables are handled in ReloadingClientFactory.
2. just returning a List is fine: we know the original list of URLs does not change, so just using the integer to index into that, and creating a Channel per host as I've done, does the right thing? We can just put put the javadoc as you described, to explain that it should be in the same order as received and call it done?

Reply:
Unless you want to push the refreshable down to this layer? I'd be down for that, but definitely not in this piece of work.

Reply:
I think the channel instances need to change if the server requests a different node selection strategy, that way we can update balanced channel scores based on host-targeted requests.